packages/base-driver/lib/basedriver/capabilities.js

```diff
@@ -336,43 +335,74 @@ function processCapabilities(
+      return obj;
+    }
 
-  let caps = _.cloneDeep(originalCaps);
-  if (!_.isPlainObject(appiumOptions)) {
-    throw new errors.SessionNotCreatedError(`The ${APPIUM_OPTS_CAP} capability must be an object`);
-  }
+    if (!_.isPlainObject(appiumOptions)) {
+      throw new errors.SessionNotCreatedError(`The ${PREFIXED_APPIUM_OPTS_CAP} capability must be an object`);
+    }
+    if (_.isEmpty(appiumOptions)) {
+      return obj;
+    }
 
-  // first get rid of any prefixes inside appium:options
-  const strippedAppiumOptions = stripAppiumPrefixes(appiumOptions);
-  // warn if we are going to overwrite any keys on the base caps object
-  const overwrittenKeys = _.intersection(Object.keys(caps), Object.keys(strippedAppiumOptions));
-  if (overwrittenKeys.length > 0) {
-    log.warn(
-      `Found capabilities inside ${PREFIXED_APPIUM_OPTS_CAP} that will overwrite ` +
-        `capabilities at the top level: ${JSON.stringify(overwrittenKeys)}`
+    log.debug(
+      `Found ${PREFIXED_APPIUM_OPTS_CAP} capability present; will promote items inside to caps`
     );
-  }
 
-  // now just apply them to the main caps object
-  caps = {...caps, ...strippedAppiumOptions};
+    const shouldAddVendorPrefix = (/** @type {string} */ capName) => !capName.startsWith(APPIUM_VENDOR_PREFIX);
+    const verifyIfAcceptable = (capName) => {
+      if (!_.isString(capName)) {
+        throw new errors.SessionNotCreatedError(
+          `Capability names in ${PREFIXED_APPIUM_OPTS_CAP} must be strings. '${capName}' is unexpected`
+        );
+      }
+      if (isStandardCap(capName)) {
+        throw new errors.SessionNotCreatedError(
+          `${PREFIXED_APPIUM_OPTS_CAP} must only contain vendor-specific capabilties. '${capName}' is unexpected`
+        );
+      }
+      return capName;
+    };
+    const preprocessedOptions = _.toPairs(_.cloneDeep(appiumOptions))
+      .map(([name, value]) => [verifyIfAcceptable(name), value])
+      .map(([name, value]) => [shouldAddVendorPrefix(name) ? `${APPIUM_VENDOR_PREFIX}${name}` : name, value])
+      .reduce((acc, [name, value]) => {acc[name] = value; return acc;}, {});
+    // warn if we are going to overwrite any keys on the base caps object
+    const overwrittenKeys = _.intersection(Object.keys(obj), Object.keys(preprocessedOptions));
+    if (overwrittenKeys.length > 0) {
+      log.warn(
+        `Found capabilities inside ${PREFIXED_APPIUM_OPTS_CAP} that will overwrite ` +
+          `capabilities at the top level: ${JSON.stringify(overwrittenKeys)}`
+      );
+    }
+    const res = _.cloneDeep(obj);
+    delete res[PREFIXED_APPIUM_OPTS_CAP];
+    return {...res, ...preprocessedOptions};
+  };
 
-  // and remove all traces of the options cap
-  delete caps[APPIUM_OPTS_CAP];
-  return caps;
+  const {alwaysMatch, firstMatch} = originalCaps;
+  const result = {};
+  if (_.isPlainObject(alwaysMatch)) {
+    result.alwaysMatch = promoteForObject(alwaysMatch);
+  } else if (_.has(originalCaps, 'alwaysMatch')) {
```

Comment:
one thing to note is that TS has a tough time with `_.has()`.  a way to do this that it _does_ understand would be

```suggestion
  } else if ('alwaysMatch' in originalCaps) {
```

This is likely a non-issue in this case, but just for future reference--`_.has()` is not very useful nowadays because of nullish coalescing

Reply:
changed